# Q1

# The three measure of central tendencies are:-
# 1) Mean
# 2) median
# 3) mode

# Q2

# 1) Mean: mean gives the average value of the entire dataset. It is a good measure when there are no outliers in the entire dataset.
# 2) median: Median gives the central value after sorting the data in ascending order. If there are odd data in the dataset then the central element will be the median and if there are even entries in the dataset then the mean of central 2 values will be the median. If there are any outliers in the data it allows to find the central measure with less skewness.
# 3) Mode: Mode is given by the data which has the highest frequency in the dataset. Modes are often used in case of data which are qualitative.

# Q3

In [3]:
import numpy as np
from scipy import stats
heights=[178,177,176,177,178.2,178,175,179,180,175,178.9,176.2,177,172.5,178,176.5]
mean=np.mean(heights)
median=np.median(heights)
mode=stats.mode(heights)

/tmp/ipykernel_77/824704034.py:6: FutureWarning: Unlike other reduction functions (e.g. `skew`, `kurtosis`), the default behavior of `mode` typically preserves the axis it acts along. In SciPy 1.11.0, this behavior will change: the default value of `keepdims` will become False, the `axis` over which the statistic is taken will be eliminated, and the value None will no longer be accepted. Set `keepdims` to True or False to avoid this warning.
  mode=stats.mode(heights)


In [8]:
print("the mean height is : %f, the mdian of heights is :%f"%(mean,median), "and the mode of the heights is:",mode)

the mean height is : 177.018750, the mdian of heights is :177.000000 and the mode of the heights is: ModeResult(mode=array([177.]), count=array([3]))


# Q4

In [11]:
import numpy as np
heights=[178,177,176,177,178.2,178,175,179,180,175,178.9,176.2,177,172.5,178,176.5]
standard_deviation=np.std(heights)

In [12]:
standard_deviation

1.7885814036548633

# Q5

# Range: The range is the simplest measure of dispersion, representing the difference between the maximum and minimum values in a dataset. It gives a rough estimate of the spread but does not account for the distribution of values in between. For example, if we have a dataset of heights: 178, 185, 156, 152, 176, the range would be 185 - 152 = 33. However, the range does not provide information about how the height is distributed within this range.
# Variance: The variance measures the average squared deviation of each data point from the mean. It takes into account the entire dataset, providing a more comprehensive measure of dispersion. For the same example above asd heights The variance is calculated as:-

In [13]:
h=[178,185,156,152,176]
variance=np.var(h)
variance

168.64000000000001

# Here the variance is 168.64 which shows that the each data deviates from the mean by  square root of variance on an average.
# Standard Deviation: The standard deviation is another measure of dispersion that provides the average distance between each data point and the mean. It is the square root of the variance and is often preferred due to its easier interpretability. In the previous example, the standard deviation would be the square root of the variance, approximately 

# Q6

# A Venn diagram is a visual representation of the relationships between different sets or groups of objects or elements. It consists of overlapping circles (or other closed curves) that represent the sets, with the overlapping regions indicating the elements that belong to multiple sets. The purpose of a Venn diagram is to illustrate the logical relationships between sets, showing the shared and distinct elements.

# Q7

In [14]:
A={2,3,4,5,6,7}
B={0,2,6,8,10}
print("Intersection of A and B is:",A.intersection(B))
print("Union of A and B is:",A.union(B))

Intersection of A and B is: {2, 6}
Union of A and B is: {0, 2, 3, 4, 5, 6, 7, 8, 10}


# Q8

# The skewness in data simply means when the data is not following the normal distribution curve but because of certain outliers the normal curve gets shifted from its original position either to its right or to its left and hence giving it the names of right skewed or left skewed datasets.

# Q9

# In a right skewed dataset the median is always less than or equal to the mean and the mode is always less than or equal to the median

# Q10

# Covariance:
Covariance measures the extent to which two variables change or vary together. It quantifies the direction and strength of the linear relationship between two variables. A positive covariance indicates a positive relationship, meaning that when one variable increases, the other tends to increase as well. A negative covariance indicates a negative relationship, where one variable tends to decrease as the other increases. However, the magnitude of covariance alone does not indicate the strength of the relationship.

# Correlation:
Correlation, on the other hand, measures the strength and direction of the linear relationship between two variables, while also providing a standardized measure that can be compared. Correlation is always between -1 and 1. A correlation of 1 indicates a perfect positive linear relationship, a correlation of -1 indicates a perfect negative linear relationship, and a correlation of 0 indicates no linear relationship. Correlation is unitless, making it easier to compare across different datasets or variables. 

# Q11

# sample mean= (sum of all observations)/(number of total observations in the given sample)

In [1]:
import numpy as np
df=np.random.randint(70,120,40)
mean=np.mean(df)

In [2]:
mean

99.1

# Q12

# For normal distribution the mean, median and mode are all equal.

# Q13

# Covariance:
Covariance measures the extent to which two variables change or vary together. It quantifies the direction and strength of the linear relationship between two variables. A positive covariance indicates a positive relationship, meaning that when one variable increases, the other tends to increase as well. A negative covariance indicates a negative relationship, where one variable tends to decrease as the other increases. However, the magnitude of covariance alone does not indicate the strength of the relationship.

# Correlation:
Correlation, on the other hand, measures the strength and direction of the linear relationship between two variables, while also providing a standardized measure that can be compared. Correlation is always between -1 and 1. A correlation of 1 indicates a perfect positive linear relationship, a correlation of -1 indicates a perfect negative linear relationship, and a correlation of 0 indicates no linear relationship. Correlation is unitless, making it easier to compare across different datasets or variables.

# Q14

# Outliers have a significant effect on the mean but a less significant effect on the median, but they have a very significant effect on the measure of dispersion

In [7]:
list1=[70,80,85,90,95,100]
mean1=np.mean(list1)
median1=np.median(list1)
range_list1=np.ptp(list1)
std1=np.std(list1)

list2=[70,80,85,90,95,200]
mean2=np.mean(list2)
median2=np.median(list2)
range_list2=np.ptp(list2)
std2=np.std(list2)

In [8]:
import pandas as pd
df=pd.DataFrame({'Central_tendencies':['mean','median','range','standard_deviation'],
                'list1':[mean1,median1,range_list1,std1],
                'list2':[mean2,median2,range_list2,std2]})

In [9]:
df

Central_tendencies      list1       list2
0                mean  86.666667  103.333333
1              median  87.500000   87.500000
2               range  30.000000  130.000000
3  standard_deviation   9.860133   43.938088

# From the example above we can see that the mean of list1 and list2 differ alot but the median is not changed here.
# And also we can see that the outlier have a very significant effect on the measure of dispersion which are range and standard deviation. 